beacon_node/beacon_chain/src/beacon_chain.rs

```diff
@@ -1928,6 +1975,72 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
         Ok(())
     }
 
+    /// This function takes a configured weak subjectivity `Checkpoint` and the latest finalized `Checkpoint`.
+    /// If the weak subjectivity checkpoint and finalized checkpoint share the same epoch, we compare roots.
+    /// If we the weak subjectivity checkpoint is from an older epoch, we iterate back through roots in the
+    /// canonical chain until we reach the finalized checkpoint from the correct epoch, and compare roots.
+    /// This must called on startup, as well as any time the canonical head changes.
+    pub fn verify_weak_subjectivity_checkpoint(
+        &self,
+        finalized_checkpoint: Checkpoint,
+        wss_checkpoint: Checkpoint,
+        state: &BeaconState<T::EthSpec>,
+    ) {
+        info!(self.log, "Verifying the configured weak subjectivity checkpoint"; "weak_subjectivity_epoch" => wss_checkpoint.epoch, "weak_subjectivity_root" => format!("{:?}", wss_checkpoint.root));
+        // If epochs match, simply compare roots.
+        if wss_checkpoint.epoch == finalized_checkpoint.epoch
+            && wss_checkpoint.root != finalized_checkpoint.root
+        {
+            self.ws_failure_shutdown(finalized_checkpoint, wss_checkpoint);
+        } else if wss_checkpoint.epoch < finalized_checkpoint.epoch {
+            let slot = wss_checkpoint
+                .epoch
+                .start_slot(T::EthSpec::slots_per_epoch());
+
+            // Because `root_at_slot` uses `rev_iter_block_roots`, this iterator will start
+            // at the canonical chain head and iterate backwards. If the given slot is a skipped slot,
+            // it will return the root of the closest prior non-skipped slot.
+            match self.root_at_slot_from_state(slot, state) {
+                Ok(Some(root)) => {
+                    if root != wss_checkpoint.root {
+                        self.ws_failure_shutdown(finalized_checkpoint, wss_checkpoint);
+                    }
+                }
+                Ok(None) => {
+                    let mut shutdown_sender = self.shutdown_sender();
+                    error!(
+                        self.log,
+                        "Could not find the given weak subjectivity checkpoint root."
+                    );
+                    let _ = shutdown_sender.try_send("Weak subjectivity checkpoint verification failed.")
```

Comment:
If this sender fails for any reason we'll end up importing/running with the offending state that conflicts with WSJ.

This function (`verify_weak_subjectivity_checkpoint`) could return a `Result<(), BeaconChainError>` which upstream functions can use to either reject the block or abort client startup.

If we do that then we can do the logging in the upstream components and give a bit more details (perhaps the offending block root). Also, I think we should do these as `crit` instead of `error`. Sorry, trying not to be annoying haha!

Reply:
Ok, updated the error handling, let me know what you think. Definitely not annoying, I really appreciate the feedback!